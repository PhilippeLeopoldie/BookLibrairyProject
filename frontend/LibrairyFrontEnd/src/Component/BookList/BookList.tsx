import { Book } from "../Book/Book";
import { OpinionEdit } from "../OpinionEdit/OpinionEdit";
import { BookType } from "../../Type";
import FetchApi from "../../FetchApi";
import url from "../../Url";
import { useEffect, useState } from "react";
import React from "react";

export const BookList = () => {
  const [books, setBooks] = useState<BookType[]>([]);

  const handleDeleteBook = (bookId: number) => {
    setBooks((books) => books.filter((book) => book.book.id !== bookId));
  };

  const fetchBooks = async () => {
    try {
      const booksData = await FetchApi(url + "api/Book");
      setBooks(booksData.$values);
    } catch (error) {
      console.error("Error fetching books:", error);
    }
  };

  useEffect(() => {
    fetchBooks();
  }, []);

  const random = books?.at(Math.floor(Math.random() * books.length));
  return (
    <div className="books">
      <h2>Recommendation of the day</h2>
      <h3>{random?.book.title}</h3>
      <div className="bookcontainer">
        {books &&
          books
            .sort((a, b) => b.book.id - a.book.id)
            .map((bookDetail, index) => (
              <Book key={bookDetail.book.id} book={bookDetail.book} onDelete={handleDeleteBook} />
            ))}
      </div>
    </div>
  );
};
