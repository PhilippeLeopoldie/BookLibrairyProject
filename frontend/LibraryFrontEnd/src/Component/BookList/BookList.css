.books__header--flex {
  display:flex;
  flex-direction: column;
}

.booksContainer {
  display: grid;
  justify-content: center;
  margin: 4% 10%;
}

@media screen and (min-width: 768px) {
  .books__header--flex {
    margin-top: 4rem;
  }

  .booksContainer {
    grid-template-columns: 1fr 1fr;
    margin-bottom: 0%;
  }
}

@media screen and (min-width: 1024px) {
  .booksContainer {
    grid-template-columns: 1fr 1fr 1fr;
  }
}

@media screen and (min-width: 1458px) {
  .booksContainer {
    grid-template-columns: 1fr 1fr 1fr 1fr;
  }
}