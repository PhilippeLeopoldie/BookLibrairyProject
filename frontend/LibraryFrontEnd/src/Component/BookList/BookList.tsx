import { Book } from "../Book/Book";
import { BookType } from "../../Type";
import url from "../../Url";
import { useEffect, useState } from "react";


export const BookList = () => {
  const [books, setBooks] = useState<BookType[] | null>(null);

  const handleDeleteBook = (bookId: number) => {

    setBooks((books) => {
      if(books!==null) {
        return books.filter((book) => book.book.id !== bookId);
      } else {
        return[]
    }
  }) 

  };
    
  

  const fetchBooks = async () => {
    try {
      const booksResponse : Response = (await fetch(url + "api/Book"));
      if(booksResponse.status === 200) {
        const booksResponseData  = await booksResponse.json();
        setBooks(booksResponseData.$values);
      } else if (booksResponse.status === 404) {
        console.log(booksResponse);
      }
    } catch (error) {
      console.error("Error fetching books:", error);
    }
  };

  
  useEffect(() => {
    fetchBooks();
  }, []);

  if(!books) {
    return <h1>Loading...</h1>
  }
  

  const random = books?.at(Math.floor(Math.random() * books.length));
  return (
    <div className="books">
      <h2>Recommendation of the day</h2>
      <h3>{random?.book.title}</h3>
      <div className="booksContainer">
        {books &&
          books
            .sort((a, b) => b.book.id - a.book.id)
            .map((bookDetail, index) => (
              <Book key={bookDetail.book.id} book={bookDetail.book} onDelete={handleDeleteBook} />
            ))}
      </div>
    </div>
  );
};
