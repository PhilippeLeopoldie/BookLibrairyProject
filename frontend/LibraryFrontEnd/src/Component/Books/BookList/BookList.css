.BookList--black {
  background-color: #000000;
  .BookList {
    background-color: transparent;
    color: #ffffff;
  }
}

.BookList--black {
  background-color:  #000000;
  color: #ffffff;
}


.bookListSearchContainer--black {
  background-color: #000000;
}

.bookListSearchContainer--black > * {
  color: #ffffff;
  background-color: #000000;
}

.books__header--flex {
  display: flex;
  flex-direction: column;
  margin-top: 0rem;
  background-color: transparent;
}
.books__header--flex > *{
  background-color: transparent;
}

.Books__Loading--black {
  margin-top: 5rem;
  background-color: #000000;
  color:#ffffff;
}

.Books__Loading--natural {
  margin-top: 5rem;
  background-color: transparent;
  color: #000000;
}

.bookListContainer {
  display: flex;
  overflow-x: auto;
  -webkit-overflow-scrolling: touch;
  margin: 4% 0%;
  background-color: transparent;
}

.bookListContainer-center {
  display: flex;
  justify-content: center;
}

@media screen and (min-width: 700px) {
  .books__header--flex {
    margin-top: 4rem;
  }

  .bookListContainer, .BookList__mostPopular {
    grid-template-columns: repeat(2, auto);
    margin-bottom: 0%;
  }
}

@media screen and (min-width: 1024px) {

  .bookListContainer {
    display: grid; 
    justify-content: center;
   }
  .bookListContainer, .BookList__mostPopular {
    grid-template-columns: repeat(3, auto);
  }
}

@media screen and (min-width: 1300px) {
  .bookListContainer .BookList__mostPopular {
    grid-template-columns: repeat(4, auto);
  }
}

@media screen and (min-width: 1458px) {
  .bookListContainer, .BookList__mostPopular {
    grid-template-columns: repeat(5, auto);
  }
}
