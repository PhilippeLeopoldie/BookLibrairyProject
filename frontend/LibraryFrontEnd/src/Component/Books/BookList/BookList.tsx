import { Book } from "../Book/Book";
import "./BookList.css";
import { BOOK_URL } from "../../../Url";
import { BOOK_TOP_BOOK_URL } from "../../../Url";
import { useContext, useEffect, useState } from "react";
import { ThemeContext } from "../../App/App";

type BooksType = {
  book: {
    id: number;
    title: string;
    author: string;
    imageUrl?: string;
    averageRate: number;
    opinions?: OpinionType | null;
  };
};

type TopBookType = {
  id: number;
  title: string;
  author: string;
  imageUrl?: string;
  averageRate: number;
  opinions?: OpinionType | null;
};

type OpinionType = {
  rate: number;
  view: string;
  userName: string;
};

export const BookList = () => {
  const theme = useContext(ThemeContext);
  const [books, setBooks] = useState<BooksType[] | null>(null);
  const [topBook, setTopBook] = useState<TopBookType[] | null>(null);
  const [numberOfBooks, setNumberOfBook] = useState<number>(3);
  const fetchBooks = async () => {
    try {
      const booksResponse: Response = await fetch(BOOK_URL);
      if (booksResponse.status === 200) {
        const booksResponseData = await booksResponse.json();
        setBooks(booksResponseData.$values);
      } else if (booksResponse.status === 404) {
        console.log(booksResponse);
      }
    } catch (error) {
      console.error("Error fetching books:", error);
    }
  };

  const fetchTopBook = async () => {
    try {
      const topBookResponse: Response = await fetch(`${BOOK_TOP_BOOK_URL} ${numberOfBooks}`);
      if (topBookResponse.status === 200) {
        const topBookResponseData = await topBookResponse.json();
        setTopBook(topBookResponseData.$values);
        console.log("topBook:", topBook);
        console.log(
          "topBook title:",
          topBook?.map((book) => book.title)
        );
      } else if (topBookResponse.status === 404) {
        console.log(topBookResponse);
      }
    } catch (error) {
      console.error("Error fetching TopBook:", error);
    }
  };

  useEffect(() => {
    fetchBooks();
    fetchTopBook();
  }, []);

  if (!books) {
    return <h1 className={"Books__Loading--" + theme}>Loading...</h1>;
  }

  return (
    <>
      <header className={`BookList--${theme} books__header--flex`}>
        <h1 className="BookList">{`Top ${numberOfBooks} Most popular`}</h1>
        <div className="BookList BookList__mostPopular">
          {topBook?.map((book) => ( 
              <Book key={book.id} book={book}/>        
          ))}
        </div>
      </header>
      <h1 className={`BookList--${theme}`}>New books</h1>
      <main className="bookListContainer">
        {books &&
          Array.isArray(books) &&
          books
            .sort((a, b) => b.book.id - a.book.id)
            .map((bookDetail) => (
              <Book key={bookDetail.book.id} book={bookDetail.book} />
            ))}
      </main>
    </>
  );
};
