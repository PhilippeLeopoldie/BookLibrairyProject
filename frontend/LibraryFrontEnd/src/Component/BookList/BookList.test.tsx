import { BookList } from "./BookList";
import { act, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

global.fetch = jest.fn().mockImplementation(() =>
  Promise.resolve({
    json: () =>
      Promise.resolve([
        { book: { id: 1, title: "Book 1", author: "Author 1" } },
      ]),
    status: 200,
    hearders: new Headers(),
  })
);

test("Should render Loading message initially", async () => {
  //Arrange
 await act( async () => {
  render(<BookList />);
 }) 

  //Act
  const loadingMessage = screen.getByText("Loading...");

  //Expect
  expect(loadingMessage).toBeInTheDocument();
});
