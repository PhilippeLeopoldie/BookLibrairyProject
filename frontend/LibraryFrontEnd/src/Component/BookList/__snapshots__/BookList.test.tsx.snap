// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookList rendering snapShot: should render BookList 1`] = `
<DocumentFragment>
  <div
    class="books"
  >
    <h2>
      Recommendation of the day
    </h2>
    <h3 />
    <div
      class="booksContainer"
      data-testid="booksContainer"
    />
  </div>
</DocumentFragment>
`;
